README.md
<p align="center">
  <img src="assets/logo.svg" width="200">
</p>

<h1 align="center">Typist JSON</h1>

<p align="center">
  <a href="https://bundlephobia.com/result?p=typist-json">
    <img src="https://badgen.net/bundlephobia/minzip/typist-json" alt="minzipped size">
  </a>
  <img src="https://badgen.net/npm/types/typist-json?icon=typescript" alt="types">
  <a href="LICENSE">
    <img src="https://badgen.net/npm/license/typist-json" alt="license">
  </a>
  <a href="https://github.com/kawmra/typist-json/actions/workflows/ci.yml">
    <img src="https://github.com/kawmra/typist-json/actions/workflows/ci.yml/badge.svg" alt="ci">
  </a>
</p>

<p align="center">
  A simple runtime JSON type checker.
</p>

# Features

- **Simple**. No JSON Schema, No validation rules
- **Type-safe**. Written in TypeScript
- **Intuitive**. Familiar syntax like TypeScript interface

Typist JSON is focused on type checking, so there is no validation rules like range of numbers or length of strings.

# Install

```shell
npm install typist-json
```

**NOTE:** Require TypeScript 4.1 or higher because Typist JSON uses [`Key Remapping`](https://www.typescriptlang.org/docs/handbook/2/mapped-types.html#key-remapping-via-as) and [`Template Literal Types`](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html).

# Example

```typescript
import { j } from "typist-json";

const NameJson = j.object({
  firstname: j.string,
  lastname: j.string,
});

const UserJson = j.object({
  name: NameJson,
  age: j.number,
  "nickname?": j.string, // optional property
});

const userJson = await fetch("/api/user")
    .then(res => res.json());

if (UserJson.check(userJson)) {
  // now, the userJson is narrowed to:
  // {
  //   name: {
  //     firstname: string
  //     lastname: string
  //   }
  //   age: number
  //   nickname?: string | undefined
  // }
}
```

## Circular References

Sometimes JSON structures can form circular references.

Typist JSON can represent circular references by wrapping checkers in the arrow function.

```ts
const FileJson = j.object({
  filename: j.string,
});

const DirJson = j.object({
  dirname: j.string,
  entries: () => j.array(j.any([FileJson, DirJson])), // references itself
});

DirJson.check({
  dirname: "animals",
  entries: [
    {
      dirname: "cat",
      entries: [
        { filename: "american-shorthair.jpg" },
        { filename: "munchkin.jpg" },
        { filename: "persian.jpg" },
      ],
    },
    {
      dirname: "dog",
      entries: [
        { filename: "chihuahua.jpg" },
        { filename: "pug.jpg" },
        { filename: "shepherd.jpg" },
      ],
    },
    { filename: "turtle.jpg" },
    { filename: "shark.jpg" },
  ],
}); // true
```

# Type checkers

## Strings

```ts
j.string.check("foo"); // true
j.string.check("bar"); // true
```

## Numbers

```ts
j.number.check(42); // true
j.number.check(12.3); // true
j.number.check("100"); // false
```

## Booleans

```ts
j.boolean.check(true); // true
j.boolean.check(false); // true
```

## Literals

```ts
j.literal("foo").check("foo"); // true
j.literal("foo").check("fooooo"); // false
```

## Arrays

```ts
j.array(j.string).check(["foo", "bar"]); // true
j.array(j.string).check(["foo", 42]); // false
j.array(j.string).check([]); // true
j.array(j.number).check([]); // true
```

## Objects

```ts
j.object({
  name: j.string,
  age: j.number,
  "nickname?": j.string,
}).check({
  name: "John",
  age: 20,
  nickname: "Johnny",
}); // true

j.object({
  name: j.string,
  age: j.number,
  "nickname?": j.string,
}).check({
  name: "Emma",
  age: 20,
}); // true, since "nickname" is optional

j.object({
  name: j.string,
  age: j.number,
  "nickname?": j.string,
}).check({
  id: "xxxx",
  type: "android",
}); // false, since "name" and "age" is required
```

If a property that ends with `?` is not optional, you should replace all trailing `?` by `??`.

<details>
<summary>More details about escaping</summary>

As mentioned above, you need to escape all trailing `?` as `??`.

```ts
j.object({
    "foo??": j.boolean,
}).check({
    "foo?": true,
}); // true
```

So if you want optional property with a name `"foo???"`,
you should use `"foo???????"` as key.

```ts
j.object({
  "foo???????": j.boolean,
}).check({}); // true, since "foo???" is optional
```
</details>

## Nulls

```ts
j.nil.check(null); // true
j.nil.check(undefined); // false
```

## Nullables

```ts
j.nullable(j.string).check("foo"); // true
j.nullable(j.string).check(null); // true
j.nullable(j.string).check(undefined); // false
```

## Unknowns

```ts
j.unknown.check("foo"); // true
j.unknown.check(123); // true
j.unknown.check(null); // true
j.unknown.check(undefined); // true
j.unknown.check([{}, 123, false, "foo"]); // true
```

## Unions

```ts
j.any([j.string, j.boolean]).check(false); // true

j.any([
  j.literal("foo"),
  j.literal("bar"),
]).check("foo"); // true
```

# Get JSON type of checkers

```ts
import { j, JsonTypeOf } from "typist-json";

const UserJson = j.object({
  name: j.string,
  age: j.number,
  "nickname?": j.string,
});

type UserJsonType = JsonTypeOf<typeof UserJson>;
// 
// ^ This is same as:
// 
// type UserJsonType = {
//   name: string;
//   age: number;
//   nickname?: string;
// }
```


CHANGELOG.md
# CHANGELOG

## 1.0.1 (2021-06-12)

- Updated README.md

## 1.0.0 (2021-06-01)

- Released


test/checker.test.ts
import {
  any,
  array,
  boolean,
  isOptionalProperty,
  literal,
  nil,
  nullable,
  number,
  object,
  string,
  unescapePropertyName,
  unknown,
} from '../src/checker';
import {Checker} from '../src';

export class TestChecker implements Checker<unknown> {
  isCalled = false;
  private readonly result: boolean;

  constructor(result: boolean) {
    this.result = result;
  }

  check(value: unknown): value is unknown {
    this.isCalled = true;
    return this.result;
  }
}

export class TestSequenceChecker implements Checker<unknown> {
  private callCount = 0;
  private readonly results: boolean[];

  constructor(results: boolean[]) {
    this.results = results;
  }

  check(value: unknown): value is unknown {
    return this.results[this.callCount++];
  }
}

describe('checker', () => {
  test('string', () => {
    expect(string.check('foo')).toBe(true);
    expect(string.check(123)).toBe(false);
    expect(string.check({})).toBe(false);
    expect(string.check(null)).toBe(false);
  });

  test('number', () => {
    expect(number.check(123)).toBe(true);
    expect(number.check(Number.POSITIVE_INFINITY)).toBe(true);
    expect(number.check(Number.NEGATIVE_INFINITY)).toBe(true);
    expect(number.check(Number.NaN)).toBe(true);
    expect(number.check('123')).toBe(false);
    expect(number.check({})).toBe(false);
    expect(number.check(null)).toBe(false);
  });

  test('boolean', () => {
    expect(boolean.check(true)).toBe(true);
    expect(boolean.check(false)).toBe(true);
    expect(boolean.check('true')).toBe(false);
    expect(boolean.check(1)).toBe(false);
    expect(boolean.check({})).toBe(false);
    expect(boolean.check([])).toBe(false);
    expect(boolean.check(null)).toBe(false);
  });

  test('nil', () => {
    expect(nil.check(null)).toBe(true);
    expect(nil.check('null')).toBe(false);
    expect(nil.check(Number.NaN)).toBe(false);
    expect(nil.check({})).toBe(false);
    expect(nil.check([])).toBe(false);
    expect(nil.check(undefined)).toBe(false);
  });

  test('unknown', () => {
    expect(unknown.check('foo')).toBe(true);
    expect(unknown.check(123)).toBe(true);
    expect(unknown.check({})).toBe(true);
    expect(unknown.check([])).toBe(true);
    expect(unknown.check(true)).toBe(true);
    expect(unknown.check(false)).toBe(true);
    expect(unknown.check(null)).toBe(true);
  });

  test('literal', () => {
    expect(literal('foo').check('foo')).toBe(true);
    expect(literal('foo').check('fooo')).toBe(false);
  });

  describe('any', () => {
    it('should return true if any of given checkers return true', () => {
      const checker1 = new TestChecker(true);
      const checker2 = new TestChecker(false);
      const checker3 = new TestChecker(false);
      expect(any([checker1, checker2]).check(null)).toBe(true);
      expect(any([checker1, checker3]).check(null)).toBe(true);
      expect(any([checker2, checker3]).check(null)).toBe(false);
    });
  });

  describe('nullable', () => {
    it('should return true if the value to be checked is null', () => {
      const checker = new TestChecker(false);
      expect(nullable(checker).check(null)).toBe(true);
    });

    it('should return true if the given checker returns true', () => {
      const checker = new TestChecker(true);
      expect(nullable(checker).check('any values')).toBe(true);
      expect(checker.isCalled).toBe(true);
    });
  });

  describe('array', () => {
    it('should return true if all elements are valid type', () => {
      const checker = new TestSequenceChecker([true, true, true, true, true]);
      expect(array(checker).check(Array(5).fill(null))).toBe(true);
    });

    it('should return false if even one element is wrong type', () => {
      const checker = new TestSequenceChecker([true, false, true]);
      expect(array(checker).check(Array(3).fill(null))).toBe(false);
    });
  });

  describe('object', () => {
    it('should return true if all properties are valid type', () => {
      expect(
        object({
          foo: new TestChecker(true),
          bar: new TestChecker(true),
        }).check({foo: null, bar: null})
      ).toBe(true);
    });

    it('should return false if even one property is wrong type', () => {
      expect(
        object({
          foo: new TestChecker(true),
          bar: new TestChecker(false),
        }).check({
          foo: null,
          bar: null,
        })
      ).toBe(false);
    });

    it('should return true if lazy checker returns true', () => {
      expect(
        object({
          foo: () => new TestChecker(true),
        }).check({foo: null})
      ).toBe(true);
    });

    it('should return false if lazy checker returns false', () => {
      expect(
        object({
          foo: () => new TestChecker(false),
        }).check({foo: null})
      ).toBe(false);
    });

    it('should return false if function does not return checker', () => {
      expect(
        object({
          foo: () => 'foo',
        }).check({foo: 'foo'})
      ).toBe(false);
      expect(
        object({
          foo: () => null,
        }).check({foo: null})
      ).toBe(false);
    });

    it('should return false if required property is missing', () => {
      expect(object({foo: new TestChecker(true)}).check({})).toBe(false);
      expect(object({'bar??': new TestChecker(true)}).check({})).toBe(false);
      expect(object({'foobar????????': new TestChecker(true)}).check({})).toBe(
        false
      );
    });

    it('should return true if optional property is missing', () => {
      expect(
        object({
          'foo?': new TestChecker(false),
          'bar???': new TestChecker(false),
          'foobar???????': new TestChecker(false),
        }).check({})
      ).toBe(true);
    });
  });
});

describe('isOptionalProperty', () => {
  test('empty', () => {
    expect(isOptionalProperty('')).toBe(false);
    expect(isOptionalProperty('?')).toBe(true);
    expect(isOptionalProperty('??')).toBe(false);
    expect(isOptionalProperty('???')).toBe(true);
    expect(isOptionalProperty('????')).toBe(false);
    expect(isOptionalProperty('?????')).toBe(true);
  });

  test('name', () => {
    expect(isOptionalProperty('foo')).toBe(false);
    expect(isOptionalProperty('foo?')).toBe(true);
    expect(isOptionalProperty('foo??')).toBe(false);
    expect(isOptionalProperty('foo???')).toBe(true);
    expect(isOptionalProperty('foo????')).toBe(false);
    expect(isOptionalProperty('foo?????')).toBe(true);
  });

  test('name with question mark', () => {
    expect(isOptionalProperty('foo?bar')).toBe(false);
    expect(isOptionalProperty('foo?bar?')).toBe(true);
    expect(isOptionalProperty('foo?bar??')).toBe(false);
    expect(isOptionalProperty('foo?bar???')).toBe(true);
    expect(isOptionalProperty('foo?bar????')).toBe(false);
    expect(isOptionalProperty('foo?bar?????')).toBe(true);
  });
});

describe('unescapePropertyName', () => {
  test('empty', () => {
    expect(unescapePropertyName('')).toBe('');
    expect(unescapePropertyName('?')).toBe('');
    expect(unescapePropertyName('??')).toBe('?');
    expect(unescapePropertyName('???')).toBe('?');
    expect(unescapePropertyName('????')).toBe('??');
    expect(unescapePropertyName('?????')).toBe('??');
  });

  test('name', () => {
    expect(unescapePropertyName('foo')).toBe('foo');
    expect(unescapePropertyName('foo?')).toBe('foo');
    expect(unescapePropertyName('foo??')).toBe('foo?');
    expect(unescapePropertyName('foo???')).toBe('foo?');
    expect(unescapePropertyName('foo????')).toBe('foo??');
    expect(unescapePropertyName('foo?????')).toBe('foo??');
  });

  test('name with question mark', () => {
    expect(unescapePropertyName('foo?bar')).toBe('foo?bar');
    expect(unescapePropertyName('foo?bar?')).toBe('foo?bar');
    expect(unescapePropertyName('foo?bar??')).toBe('foo?bar?');
    expect(unescapePropertyName('foo?bar???')).toBe('foo?bar?');
    expect(unescapePropertyName('foo?bar????')).toBe('foo?bar??');
    expect(unescapePropertyName('foo?bar?????')).toBe('foo?bar??');
  });
});


test/test-checker.ts
import {Checker} from '../src';

export class TestChecker implements Checker<unknown> {
  isCalled = false;
  private readonly result: boolean;

  constructor(result: boolean) {
    this.result = result;
  }

  check(value: unknown): value is unknown {
    this.isCalled = true;
    return this.result;
  }
}

export class TestSequenceChecker implements Checker<unknown> {
  private callCount = 0;
  private readonly results: boolean[];

  constructor(results: boolean[]) {
    this.results = results;
  }

  check(value: unknown): value is unknown {
    return this.results[this.callCount++];
  }
}


test/validator.test.ts
import {
  any,
  array,
  boolean,
  isOptionalProperty,
  literal,
  nil,
  nullable,
  number,
  object,
  string,
  unescapePropertyName,
  unknown,
} from '../src/validator';
import {TestSequenceValidator, TestValidator} from './test-validators';

describe('validator', () => {
  test('string', () => {
    expect(string.validate('foo')).toBe(true);
    expect(string.validate(123)).toBe(false);
    expect(string.validate({})).toBe(false);
    expect(string.validate(null)).toBe(false);
  });

  test('number', () => {
    expect(number.validate(123)).toBe(true);
    expect(number.validate(Number.POSITIVE_INFINITY)).toBe(true);
    expect(number.validate(Number.NEGATIVE_INFINITY)).toBe(true);
    expect(number.validate(Number.NaN)).toBe(true);
    expect(number.validate('123')).toBe(false);
    expect(number.validate({})).toBe(false);
    expect(number.validate(null)).toBe(false);
  });

  test('boolean', () => {
    expect(boolean.validate(true)).toBe(true);
    expect(boolean.validate(false)).toBe(true);
    expect(boolean.validate('true')).toBe(false);
    expect(boolean.validate(1)).toBe(false);
    expect(boolean.validate({})).toBe(false);
    expect(boolean.validate([])).toBe(false);
    expect(boolean.validate(null)).toBe(false);
  });

  test('nil', () => {
    expect(nil.validate(null)).toBe(true);
    expect(nil.validate('null')).toBe(false);
    expect(nil.validate(Number.NaN)).toBe(false);
    expect(nil.validate({})).toBe(false);
    expect(nil.validate([])).toBe(false);
    expect(nil.validate(undefined)).toBe(false);
  });

  test('unknown', () => {
    expect(unknown.validate('foo')).toBe(true);
    expect(unknown.validate(123)).toBe(true);
    expect(unknown.validate({})).toBe(true);
    expect(unknown.validate([])).toBe(true);
    expect(unknown.validate(true)).toBe(true);
    expect(unknown.validate(false)).toBe(true);
    expect(unknown.validate(null)).toBe(true);
  });

  test('literal', () => {
    expect(literal('foo').validate('foo')).toBe(true);
    expect(literal('foo').validate('fooo')).toBe(false);
  });

  describe('any', () => {
    it('should return true if any of given validators return true', () => {
      const validator1 = new TestValidator(true);
      const validator2 = new TestValidator(false);
      const validator3 = new TestValidator(false);
      expect(any([validator1, validator2]).validate(null)).toBe(true);
      expect(any([validator1, validator3]).validate(null)).toBe(true);
      expect(any([validator2, validator3]).validate(null)).toBe(false);
    });
  });

  describe('nullable', () => {
    it('should return true if the value to be validated is null', () => {
      const validator = new TestValidator(false);
      expect(nullable(validator).validate(null)).toBe(true);
    });

    it('should return true if the given validator returns true', () => {
      const validator = new TestValidator(true);
      expect(nullable(validator).validate('any values')).toBe(true);
      expect(validator.isCalled).toBe(true);
    });
  });

  describe('array', () => {
    it('should return true if all elements are valid', () => {
      const validator = new TestSequenceValidator([
        true,
        true,
        true,
        true,
        true,
      ]);
      expect(array(validator).validate(Array(5).fill(null))).toBe(true);
    });

    it('should return false if even one element is invalid', () => {
      const validator = new TestSequenceValidator([true, false, true]);
      expect(array(validator).validate(Array(3).fill(null))).toBe(false);
    });
  });

  describe('object', () => {
    it('should return true if all properties are valid', () => {
      expect(
        object({
          foo: new TestValidator(true),
          bar: new TestValidator(true),
        }).validate({foo: null, bar: null})
      ).toBe(true);
    });

    it('should return false if even one property is invalid', () => {
      expect(
        object({
          foo: new TestValidator(true),
          bar: new TestValidator(false),
        }).validate({
          foo: null,
          bar: null,
        })
      ).toBe(false);
    });

    it('should return false if required property is missing', () => {
      expect(object({foo: new TestValidator(true)}).validate({})).toBe(false);
      expect(object({'bar??': new TestValidator(true)}).validate({})).toBe(
        false
      );
      expect(
        object({'foobar????????': new TestValidator(true)}).validate({})
      ).toBe(false);
    });

    it('should return true if optional property is missing', () => {
      expect(
        object({
          'foo?': new TestValidator(false),
          'bar???': new TestValidator(false),
          'foobar???????': new TestValidator(false),
        }).validate({})
      ).toBe(true);
    });
  });
});

describe('isOptionalProperty', () => {
  test('empty', () => {
    expect(isOptionalProperty('')).toBe(false);
    expect(isOptionalProperty('?')).toBe(true);
    expect(isOptionalProperty('??')).toBe(false);
    expect(isOptionalProperty('???')).toBe(true);
    expect(isOptionalProperty('????')).toBe(false);
    expect(isOptionalProperty('?????')).toBe(true);
  });

  test('name', () => {
    expect(isOptionalProperty('foo')).toBe(false);
    expect(isOptionalProperty('foo?')).toBe(true);
    expect(isOptionalProperty('foo??')).toBe(false);
    expect(isOptionalProperty('foo???')).toBe(true);
    expect(isOptionalProperty('foo????')).toBe(false);
    expect(isOptionalProperty('foo?????')).toBe(true);
  });

  test('name with question mark', () => {
    expect(isOptionalProperty('foo?bar')).toBe(false);
    expect(isOptionalProperty('foo?bar?')).toBe(true);
    expect(isOptionalProperty('foo?bar??')).toBe(false);
    expect(isOptionalProperty('foo?bar???')).toBe(true);
    expect(isOptionalProperty('foo?bar????')).toBe(false);
    expect(isOptionalProperty('foo?bar?????')).toBe(true);
  });
});

describe('unescapePropertyName', () => {
  test('empty', () => {
    expect(unescapePropertyName('')).toBe('');
    expect(unescapePropertyName('?')).toBe('');
    expect(unescapePropertyName('??')).toBe('?');
    expect(unescapePropertyName('???')).toBe('?');
    expect(unescapePropertyName('????')).toBe('??');
    expect(unescapePropertyName('?????')).toBe('??');
  });

  test('name', () => {
    expect(unescapePropertyName('foo')).toBe('foo');
    expect(unescapePropertyName('foo?')).toBe('foo');
    expect(unescapePropertyName('foo??')).toBe('foo?');
    expect(unescapePropertyName('foo???')).toBe('foo?');
    expect(unescapePropertyName('foo????')).toBe('foo??');
    expect(unescapePropertyName('foo?????')).toBe('foo??');
  });

  test('name with question mark', () => {
    expect(unescapePropertyName('foo?bar')).toBe('foo?bar');
    expect(unescapePropertyName('foo?bar?')).toBe('foo?bar');
    expect(unescapePropertyName('foo?bar??')).toBe('foo?bar?');
    expect(unescapePropertyName('foo?bar???')).toBe('foo?bar?');
    expect(unescapePropertyName('foo?bar????')).toBe('foo?bar??');
    expect(unescapePropertyName('foo?bar?????')).toBe('foo?bar??');
  });
});


src/validator.ts
export interface Validator<T> {
  validate(value: unknown): value is T;
}

export type JsonOf<T extends Validator<unknown>> = T extends Validator<infer U>
  ? U
  : never;

export const string: Validator<string> = {
  validate: (value: unknown): value is string => {
    return typeof value === 'string';
  },
};

export const number: Validator<number> = {
  validate: (value: unknown): value is number => {
    return typeof value === 'number';
  },
};

export const boolean: Validator<boolean> = {
  validate: (value: unknown): value is boolean => {
    return typeof value === 'boolean';
  },
};

export const nil: Validator<null> = {
  validate: (value: unknown): value is null => {
    return value === null;
  },
};

export const unknown: Validator<unknown> = {
  validate: (value: unknown): value is unknown => {
    return true;
  },
};

export function literal<T extends string>(str: T): Validator<T> {
  return {
    validate(value: unknown): value is T {
      return value === str;
    },
  };
}

export function any<T extends Validator<unknown>>(
  validators: T[]
): Validator<JsonOf<T>> {
  return {
    validate(value: unknown): value is JsonOf<T> {
      return validators.some(validator => validator.validate(value));
    },
  };
}

export function nullable<T>(validator: Validator<T>): Validator<T | null> {
  return {
    validate(value: unknown): value is T | null {
      return value === null || validator.validate(value);
    },
  };
}

export function array<T extends Validator<unknown>>(
  validator: T
): Validator<JsonOf<T>[]> {
  return {
    validate(value: unknown): value is JsonOf<T>[] {
      return Array.isArray(value) && value.every(it => validator.validate(it));
    },
  };
}

export function object<T extends {[key: string]: Validator<unknown>}>(
  properties: T
): Validator<ExpandRecursively<ObjectJsonOf<T>>> {
  return {
    validate(value: unknown): value is ExpandRecursively<ObjectJsonOf<T>> {
      if (typeof value !== 'object' || value === null) return false;
      return Object.keys(properties).every(key => {
        const unescapedKey = unescapePropertyName(key);
        if (has(value, unescapedKey)) {
          return properties[key].validate(value[unescapedKey]);
        } else {
          return isOptionalProperty(key);
        }
      });
    },
  };
}

type FilterOptionalPropertyName<T> = T extends `${infer U}??`
  ? `${FilterOptionalPropertyName<U>}?`
  : T extends `${infer U}?`
  ? U
  : never;

type FilterRequiredPropertyName<T> = T extends `${infer U}??`
  ? `${FilterRequiredPropertyName<U>}?`
  : T extends `${infer U}?`
  ? never
  : T;

type ObjectJsonOf<T extends {[k: string]: Validator<unknown>}> = {
  [K in keyof T as FilterRequiredPropertyName<K>]-?: JsonOf<T[K]>;
} &
  {[K in keyof T as FilterOptionalPropertyName<K>]+?: JsonOf<T[K]>};

type ExpandRecursively<T> = T extends object
  ? T extends infer O
    ? {[K in keyof O]: ExpandRecursively<O[K]>}
    : never
  : T;

export function isOptionalProperty(propertyName: string): boolean {
  return propertyName.match(/[^?]?(?:\?\?)*(\?)?$/)?.[1] === '?';
}

export function unescapePropertyName(propertyName: string): string {
  const match = propertyName.match(/^(.*?[^?]?)((?:\?\?)*)\??$/);
  if (match === null) {
    return propertyName;
  }
  const head = match[1] ?? '';
  const questionMarks = match[2].replace(/\?\?/g, '?') ?? '';
  return `${head}${questionMarks}`;
}

// https://github.com/microsoft/TypeScript/issues/21732#issuecomment-663994772
function has<P extends PropertyKey>(
  target: object,
  property: P
): target is {[K in P]: unknown} {
  return property in target;
}


test-d/index.test-d.ts
import {expectNotType, expectType} from 'tsd';
import {Checker, j} from '../lib';

// string
expectType<Checker<string>>(j.string);

// number
expectType<Checker<number>>(j.number);

// boolean
expectType<Checker<boolean>>(j.boolean);

// null
expectType<Checker<null>>(j.nil);

// unknown
expectType<Checker<unknown>>(j.unknown);

// literal
expectType<Checker<'literal'>>(j.literal('literal'));

// any
expectType<
  Checker<
    | string
    | number
    | boolean
    | null
    | 'literal'
    | (number | boolean)
    | string[]
    | {foo: string}
  >
>(
  j.any([
    j.string,
    j.number,
    j.boolean,
    j.nil,
    j.literal('literal'),
    j.any([j.number, j.boolean]),
    j.array(j.string),
    j.object({foo: j.string}),
  ])
);
expectType<Checker<unknown>>(j.any([j.string, j.unknown, j.number]));

// nullable
expectType<Checker<string | null>>(j.nullable(j.string));
expectType<Checker<string[] | null>>(j.nullable(j.array(j.string)));
expectType<Checker<{foo: string} | null>>(
  j.nullable(j.object({foo: j.string}))
);
expectType<Checker<unknown>>(j.nullable(j.unknown));

// array
expectType<Checker<string[]>>(j.array(j.string));
expectType<Checker<number[]>>(j.array(j.number));
expectType<Checker<boolean[]>>(j.array(j.boolean));
expectType<Checker<null[]>>(j.array(j.nil));
expectType<Checker<unknown[]>>(j.array(j.unknown));
expectType<Checker<'literal'[]>>(j.array(j.literal('literal')));
expectType<Checker<(string | number)[]>>(j.array(j.any([j.string, j.number])));
expectType<Checker<string[][]>>(j.array(j.array(j.string)));
expectType<Checker<{foo: string}[]>>(j.array(j.object({foo: j.string})));

// object
expectType<Checker<{}>>(j.object({}));
expectType<Checker<{foo: string}>>(j.object({foo: () => j.string}));
expectType<Checker<{foo: never}>>(j.object({foo: () => 'foo'}));
expectType<Checker<{optional?: string}>>(j.object({'optional?': j.string}));
expectNotType<Checker<{optional: string | undefined}>>(
  j.object({'optional?': j.string})
);
expectType<Checker<{'escaped?': string}>>(j.object({'escaped??': j.string}));
expectType<Checker<{'escaped_optional?'?: string}>>(
  j.object({'escaped_optional???': j.string})
);
expectNotType<Checker<{'escaped_optional?': string | undefined}>>(
  j.object({'escaped_optional???': j.string})
);
expectType<
  Checker<{
    string: string;
    number: number;
    boolean: boolean;
    nil: null;
    unknown: unknown;
    literal: 'literal';
    union: string | number;
    nullable: string | null;
    array: string[];
    nested: {
      string: string;
    };
    optional?: string;
    'escaped?': string;
    'escaped_optional?'?: string;
    lazy: number | string;
  }>
>(
  j.object({
    string: j.string,
    number: j.number,
    boolean: j.boolean,
    nil: j.nil,
    unknown: j.unknown,
    literal: j.literal('literal'),
    union: j.any([j.string, j.number]),
    nullable: j.nullable(j.string),
    array: j.array(j.string),
    nested: j.object({
      string: j.string,
    }),
    'optional?': j.string,
    'escaped??': j.string,
    'escaped_optional???': j.string,
    lazy: () => j.any([j.number, j.string]),
  })
);


src/checker.ts
import {Checker, JsonTypeOf} from './types';

export const string: Checker<string> = {
  check: (value: unknown): value is string => {
    return typeof value === 'string';
  },
};

export const number: Checker<number> = {
  check: (value: unknown): value is number => {
    return typeof value === 'number';
  },
};

export const boolean: Checker<boolean> = {
  check: (value: unknown): value is boolean => {
    return typeof value === 'boolean';
  },
};

export const nil: Checker<null> = {
  check: (value: unknown): value is null => {
    return value === null;
  },
};

export const unknown: Checker<unknown> = {
  check: (value: unknown): value is unknown => {
    return true;
  },
};

export function literal<T extends string>(str: T): Checker<T> {
  return {
    check(value: unknown): value is T {
      return value === str;
    },
  };
}

export function any<T extends Checker<unknown>>(
  checkers: T[]
): Checker<JsonTypeOf<T>> {
  return {
    check(value: unknown): value is JsonTypeOf<T> {
      return checkers.some(checker => checker.check(value));
    },
  };
}

export function nullable<T>(checker: Checker<T>): Checker<T | null> {
  return {
    check(value: unknown): value is T | null {
      return value === null || checker.check(value);
    },
  };
}

export function array<T extends Checker<unknown>>(
  checker: T
): Checker<JsonTypeOf<T>[]> {
  return {
    check(value: unknown): value is JsonTypeOf<T>[] {
      return Array.isArray(value) && value.every(it => checker.check(it));
    },
  };
}

export function object<T extends {[key: string]: Checker<unknown> | Function}>(
  properties: T
): Checker<ExpandRecursively<ObjectJsonTypeOf<T>>> {
  return {
    check(value: unknown): value is ExpandRecursively<ObjectJsonTypeOf<T>> {
      if (typeof value !== 'object' || value === null) return false;
      return Object.keys(properties).every(key => {
        const unescapedKey = unescapePropertyName(key);
        if (has(value, unescapedKey)) {
          const checkerOrFunction = properties[key];
          if (isChecker(checkerOrFunction)) {
            return checkerOrFunction.check(value[unescapedKey]);
          }
          const lazyChecker: unknown = checkerOrFunction();
          if (isChecker(lazyChecker)) {
            return lazyChecker.check(value[unescapedKey]);
          } else {
            return false;
          }
        } else {
          return isOptionalProperty(key);
        }
      });
    },
  };
}

function isChecker(target: unknown): target is Checker<unknown> {
  return (
    typeof target === 'object' &&
    target !== null &&
    has(target, 'check') &&
    typeof target.check === 'function'
  );
}

type FilterOptionalPropertyName<T> = T extends `${infer U}??`
  ? `${FilterOptionalPropertyName<U>}?`
  : T extends `${infer U}?`
  ? U
  : never;

type FilterRequiredPropertyName<T> = T extends `${infer U}??`
  ? `${FilterRequiredPropertyName<U>}?`
  : T extends `${infer U}?`
  ? never
  : T;

type ObjectJsonTypeOf<T extends {[k: string]: Checker<unknown> | Function}> = {
  [K in keyof T as FilterRequiredPropertyName<K>]-?: T[K] extends Checker<unknown>
    ? JsonTypeOf<T[K]>
    : T[K] extends Function
    ? LazyChecker<T[K]>
    : never;
} &
  {
    [K in keyof T as FilterOptionalPropertyName<K>]+?: T[K] extends Checker<unknown>
      ? JsonTypeOf<T[K]>
      : T[K] extends Function
      ? LazyChecker<T[K]>
      : never;
  };

type LazyChecker<T extends Function> = T extends () => Checker<infer U>
  ? U
  : never;

type ExpandRecursively<T> = T extends object
  ? T extends infer O
    ? {[K in keyof O]: ExpandRecursively<O[K]>}
    : never
  : T;

export function isOptionalProperty(propertyName: string): boolean {
  return propertyName.match(/[^?]?(?:\?\?)*(\?)?$/)?.[1] === '?';
}

export function unescapePropertyName(propertyName: string): string {
  const match = propertyName.match(/^(.*?[^?]?)((?:\?\?)*)\??$/);
  if (match === null) {
    return propertyName;
  }
  const head = match[1] ?? '';
  const questionMarks = match[2].replace(/\?\?/g, '?') ?? '';
  return `${head}${questionMarks}`;
}

// https://github.com/microsoft/TypeScript/issues/21732#issuecomment-663994772
function has<P extends PropertyKey>(
  target: object,
  property: P
): target is {[K in P]: unknown} {
  return property in target;
}
